Guards page: Copyediting the whole page (#9500)
elixir-lang/elixir: lib/elixir/pages/Guards.md

You can find the built-in list of guards [in the `Kernel` module](Kernel.html#guards). Here is an overview:

  * comparison operators ([`==`](`==/2`), [`!=`](`!=/2`), [`===`](`===/2`), [`!==`](`!==/2`),

>     [`<`](`</2`), [`<=`](`<=/2`), [`>`](`>/2`), [`>=`](`>=/2`))

Author: follows the logical order `<, >`

  * function clauses:

    ```elixir
    def type(term) when is_integer(term), do: term

>     def type(term) when is_float(term), do: round(term)

Reviewer: It should return the atoms `:integer` and `:float` 🤓

Author: oops. fixed

>     def format(term) when is_float(term), do: round(term)

Reviewer: I don't think `format` makes any more sense than `foo`.

Author: maybe `format_number` ?

Reviewer 2: Let's rewrite it to:

```elixir
def type(term) when is_integer(term), do: :integer
def type(term) when is_float(term), do: :float
```

?

Author: done

## Failing guards

> As mentioned before, a function clause will be executed if and only if its guard expression evaluates to `true`. If any other value is returned, the function clause will be skipped. In particular, guards have no concept of "truthy" or "falsey". In a nutshell, when it comes to guards, *it's `true` or nothing*.

Reviewer 2: "As mentioned before" is usually an anti-pattern.

1. Or the information is worth being repeated, which means the "as mentioned before" is irrelevant
2. Or the information is not relevant, which means it should not be repeated at all

Reviewer 2: I am not sure adding "In a nutshell, when it comes to guards, *it's `true` or nothing*." helps. At this point, we are saying the same thing 4 times:

  1. if and only if its guard expression evaluates to `true`
  2. If any other value is returned, the function clause will be skipped (same thing, but inverted)
  3. In particular, guards have no concept of "truthy" or "falsey" (can be derived from 1)
  4. In a nutshell, when it comes to guards, *it's `true` or nothing*. (same as 1)

Please remove. :)

Author: I reverted the changes, but I removed a redundant "only"

#=> false
```

> Even though the head of the list is not `nil`, the first clause for `not_nil_head?/1` fails because the expression is evaluated to `"some_value"` and not to `true` as guards are expected to return, therefore triggering the second clause which returns `false`. To make the guard behave correctly, you must ensure that the guard evaluates to `true`, like so:

Reviewer 2: The new sentence is too long.

```suggestion
Even though the head of the list is not `nil`, the first clause for `not_nil_head?/1` fails because the expression does not evaluate to `true`, but to `"some_value"`, therefore triggering the second clause which returns `false`. To make the guard behave correctly, you must ensure that the guard evaluates to `true`, like so:
```

Author: done

  def empty?(_val), do: false
end

Check.empty?(%{})
#=> true

Check.empty?({})
#=> false # true was expected!
```

This could be corrected by ensuring that no exception is raised, either via type checks like `is_map(val) and map_size(val) == 0`, or by checking equality instead, like `val == %{}`.

> It could also be corrected by using multiple guards, so that if an exception causes one guard to fail, the next one is still evaluated.

Reviewer 2: I am not sure "still" is correct, because the next one is not evaluated if the first one succeeds.

Author: I think it is correct, because it is preceded by "if an exception ...", contrasting what happens when no multiple-clauses are used

  def empty?(_val), do: false
end

Check.empty?(%{})
#=> true

Check.empty?({})
#=> true
```

## Defining custom guard expressions

> As mentioned before, only the functions and operators listed in this page are allowed in guards. However, we can take advantage of macros to write custom guards that can simplify our programs or make them more domain-specific. At the end of the day, what matters is that the *output* of the macros is boiled down to a combination of allowed expressions.

Reviewer 2: "expressions" is better than functions and operators, as some of them may be macros.

Also, please avoid passive voice.

Author: should we then correct the name of the title of the first section to "List of allowed functions, macros and operators" ?

Author: done

> Let's look at a quick case study: we want to check whether a function argument is an even or an odd integer. With pattern matching this is impossible because there is an infinite number of integers, and therefore we can't pattern match on every single one of them. So, let's focus on checking for even numbers since checking for the odd ones is almost identical.

Reviewer 2: "So, " is not necessary.

Author: removed the "So". Please have a look again, I'm trying to say "Let's do this" twice in the same paragraph.

end
```

> It would be repetitive to write every time we need this check, so, as mentioned at the beginning of this section, we can abstract this away using a macro. Remember that defining a function that performs this check wouldn't work because we can't use custom functions in guards. Instead you can use `defguard/1` and `defguardp/1` to create guard-safe macros. Here's an example how:

Reviewer 2: "guard-safe" is introduced here but it is not explained what it is anywhere else. I would revert the `-safe` addition.

Author: removed the use of guad-safe throughout the page.

end
```

> While it's possible to create custom guards with macros, it's recommended to define them using `defguard/1` and `defguardp/1` which perform additional compile-time checks and optimizations.

Reviewer: Do they perform optimizations?

Author: Yes, when they are used outside guards

Reviewer: I wouldn't call that optimizations, it is more that it keeps evaluation semantics, so I would remove that note.

Author: https://github.com/elixir-lang/elixir/blob/master/lib/elixir/lib/kernel/utils.ex#L159

Reviewer: Unquoting once per argument is about preserving evaluation semantics. Maybe I am missing something so can you link to the specific code or docs?

Author: I submitted that link before reading your comment

Reviewer: Just to clarify, it can be an optimization, but it is primarily to preserve evaluation semantics, specifically that each argument is evaluated once like expected.

Reviewer 2: Agreed with @ericmj.

Author: reverted. thank you for the discussion
